### PR TITLE
Add kyuubi-spark-extensions module

### PR DESCRIPTION
**Pull request by ulysses-you**
### _Why are the changes needed?_

The added sql module structure looks like:
```
kyuubi
  |
  - dev
      |
      - kyuubi-extension-spark_3.1
```

This PR mainly add 3 feature:
* merging small files automatically (include dynamic partition insertion case)
* insert shuffle node before Join to make AQE `OptimizeSkewedJoin` work
* stage level config isolation in AQE

Note that,  the sql rule  depend on the Apache Spark interface so we need make the sql module verion independence. Currently, this PR only supports the Spark 3.1.1.

Due to the version issue, we need to check and deploy this extension manually currently.

### _How was this patch tested?_
Add new test.

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @yaooqinn @pan3793 @turboFei

File: pom.xml [hunk 1]
```diff
@@ -35,6 +35,7 @@
         <module>kyuubi-main</module>
         <module>kyuubi-metrics</module>
         <module>kyuubi-zookeeper</module>
+        <module>externals/kyuubi-sql/kyuubi-sql-spark-3_1</module>
```
Comment (pan3793) [inline pom.xml:38]: Sort alphabetically

File: externals/kyuubi-sql/kyuubi-sql-spark-3_1/pom.xml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>kyuubi</artifactId>
+        <groupId>org.apache.kyuubi</groupId>
+        <version>1.2.0-SNAPSHOT</version>
+        <relativePath>../../../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>kyuubi-sql-spark_3.1</artifactId>
```
Comment (pan3793) [inline externals/kyuubi-sql/kyuubi-sql-spark-3_1/pom.xml:30]: the folder named `...-3_1` but artifactId is `..._3.1`, it's by design?

File: externals/kyuubi-sql/kyuubi-sql-spark-3_1/pom.xml [hunk 3]
```diff
@@ -0,0 +1,109 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (yaooqinn) [inline externals/kyuubi-sql/kyuubi-sql-spark-3_1/pom.xml:1]: dev/kyuubi-extensions-spark_3.1

**Comment by yaooqinn** [comment]
cc @cloud-fan (if interested)

**Comment by yaooqinn** [comment]
Can we add some real hammers for showing performance improvement with these extensions

File: dev/kyuubi-extension-spark_3.1/pom.xml [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>kyuubi</artifactId>
+        <groupId>org.apache.kyuubi</groupId>
+        <version>1.2.0-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>kyuubi-extension-spark_3.1</artifactId>
```
Comment (yaooqinn) [inline dev/kyuubi-extension-spark_3.1/pom.xml:30]: <name>Kyuubi Project Dev Spark Extensions</name>
    <packaging>jar</packaging>

File: dev/kyuubi-extension-spark_3.1/pom.xml [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    <properties>
+        <spark.version>3.1.1</spark.version>
+        <scala.binary.version>2.12</scala.binary.version>
```
Comment (yaooqinn) [inline dev/kyuubi-extension-spark_3.1/pom.xml:34]: unecessary

File: dev/kyuubi-extension-spark_3.1/src/main/scala/org/apache/kyuubi/sql/KyuubiSQLConf.scala [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.internal.SQLConf._
+
+object KyuubiSQLConf {
+
+  val INSERT_REPARTITION_BEFORE_WRITE =
+    buildConf("spark.sql.optimizer.insertRepartitionBeforeWrite.enabled")
+      .doc("Add repartition node at the top of plan. A approach of merging small files.")
+      .version("0.0.1")
```
Comment (yaooqinn) [inline dev/kyuubi-extension-spark_3.1/src/main/scala/org/apache/kyuubi/sql/KyuubiSQLConf.scala:28]: use 1.2.0

File: dev/kyuubi-extension-spark_3.1/pom.xml [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            <groupId>org.scalatest</groupId>
+            <artifactId>scalatest_${scala.binary.version}</artifactId>
+            <version>${scalatest.version}</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <outputDirectory>target/scala-2.12/classes</outputDirectory>
+        <testOutputDirectory>target/scala-2.12/test-classes</testOutputDirectory>
```
Comment (yaooqinn) [inline dev/kyuubi-extension-spark_3.1/pom.xml:107]: ${scala.binary.version}

File: dev/kyuubi-extension-spark_3.1/pom.xml [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-sql_${scala.binary.version}</artifactId>
+            <version>${spark.version}</version>
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
```
Comment (yaooqinn) [inline dev/kyuubi-extension-spark_3.1/pom.xml:81]: unnecessary

File: dev/kyuubi-extension-spark_3.1/pom.xml [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    <artifactId>kyuubi-extension-spark_3.1</artifactId>
+
+    <properties>
+        <spark.version>3.1.1</spark.version>
+        <scala.binary.version>2.12</scala.binary.version>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.spark</groupId>
```
Comment (yaooqinn) [inline dev/kyuubi-extension-spark_3.1/pom.xml:39]: unnecessary

**Review by yaooqinn — APPROVED** [review]
LGTM

**Comment by yaooqinn** [comment]
thanks, merged to master for v1.3.0 and branch 1.2 for v1.2.0

**Comment by pan3793** [comment]
module version of master is not correct

**Comment by yaooqinn** [comment]
nice catch can you send a followup？

**Comment by pan3793** [comment]
sure

**Comment by yangrong688** [comment]
Hi @ulysses-you, I wonder is this feature product ready? I cannot find any documentation about how to use it in Kyuubi website.

**Comment by ulysses-you** [comment]
@yangrong688 yes, it's ready.

But now this only available with Spark branch-3.1 (i.e 3.1.1 and 3.1.2). Since Kyuubi 1.2.0 has not released yet, you can download the last rc version in https://github.com/NetEase/kyuubi/releases.

About the document, yeah it's better to improve the documents about this feature. You can try this

1. move the jar which is in `extension` into `$SPARK_HOME/jars`
2. add a config into `spark-defaults.conf`: `spark.sql.extensions=org.apache.kyuubi.sql.KyuubiSparkSQLExtension`

**Comment by yangrong688** [comment]
> @yangrong688 yes, it's ready.
> 
> But now this only available with Spark branch-3.1 (i.e 3.1.1 and 3.1.2). Since Kyuubi 1.2.0 has not released yet, you can download the last rc version in https://github.com/NetEase/kyuubi/releases.
> 
> About the document, yeah it's better to improve the documents about this feature. You can try this
> 
>     1. move the jar which is in `extension` into `$SPARK_HOME/jars`
> 
>     2. add a config into `spark-defaults.conf`: `spark.sql.extensions=org.apache.kyuubi.sql.KyuubiSparkSQLExtension`

OK, thanks. I will have a try.

**Comment by yaooqinn** [comment]
please help with the doc @ulysses-you :)

**Pull request closed without merging**